### PR TITLE
Reassociation: fix counting of constant multiplicative factors

### PR DESCRIPTION
**Pull request by dneto0**
In the middle of rewriting expressions like (A*B + A*C + D) to pull common factor A out, the algorithm finds that there's actually only one A. This is unexpected, and it fires an assertion.

This can occur when A is a constant, and constant -A also appears in the terms somewhere else.

There is no harm in this situation, however, because the algorithm then creates an addition-tree, but with a single element, and that's still correct.

This bookkeeping issue was fixed later in LLVM, at https://github.com/llvm/llvm-project/commit/95abfa35d6c0b8a769399262496a3f692177ef12

Unfortunately the associated test doesn't translate cleanly to DXC-era LLVM. I've added test case reduced from our original case.

Fixed: #6829

### CODE REVIEW

File: test/HLSL/passes/reassociate/constant-counting.ll [hunk 1]
```diff
@@ -0,0 +1,189 @@
+; RUN: opt %s -reassociate -S | FileCheck %s
+
+; Issue: #6829
+
+; There are cases where Reassociation used to double-count a constant value
```
Comment (dneto0) [inline test/HLSL/passes/reassociate/constant-counting.ll:5]: I tried quite hard to reduce this test case, or to adapt the upstream test case.

Reassociation works iteratively, and it needs to generate certain intermediate results that then force the final issue.
It's very hard to follow. This is actually a cut down version of the original HLSL we got (and I put in the issue)

**Review by amaiorano — APPROVED** [review]

**Review by llvm-beanz — APPROVED** [review]
